File: Popularity/DCAFPilot/src/python/DCAF/services/popdb.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-
#pylint: disable=
"""
File       : popdb.py
Author     : Valentin Kuznetsov <vkuznet AT gmail dot com>
Description: PopDB service module
"""

# system modules
import time
import urllib
from   types import InstanceType

# package modules
import DCAF.utils.jsonwrapper as json
from DCAF.utils.sso_reader import sso_getdata
from DCAF.utils.url_utils import getdata
from DCAF.utils.utils import get_key_cert, genkey
from DCAF.services.generic import GenericService

class PopDBService(GenericService):
    """
    Helper class to provide Popularity service
    """
    def __init__(self, config=None, verbose=0):
        GenericService.__init__(self, config, verbose)
        self.name = 'popdb'
        self.url = config['services'][self.name]
        self.ckey, self.cert = get_key_cert()

    def fetch(self, api, params=None):
        "Fetch data for given api"
        url = '%s/%s?%s' % (self.url, api, urllib.urlencode(params, doseq=True))
        docid = genkey("url=%s params=%s" % (url, params))
        res = self.storage.fetch_one('cache', {'_id':docid})
        if  res and 'data' in res:
            if  self.verbose:
                print "%s::fetch url=%s, params=%s, docid=%s" \
                        % (self.name, url, params, docid)
            data = res['data']
        else:
            if  self.verbose:
                print "%s::fetch url=%s, params=%s" % (self.name, url, params)
            # NOTE: popularity DB has two different access points, one
            # within CERN network and out outside. The former does not require
            # authentication, while later passes through CERN SSO.
            # The following block reflects this, in a future, when popularity DB
            # will move into cmsweb domain we'll no longer need it
            if  self.url.find('cms-popularity-prod.cern.ch') != -1:
                data = getdata(url, ckey=self.ckey, cert=self.cert, debug=self.verbose)
            else:
                data = sso_getdata(url, ckey=self.ckey, cert=self.cert, debug=self.verbose)
            self.storage.insert('cache', {'_id':docid, 'data': data, 'url': url, 'params': params})
        data = json.loads(data)
        for row in data['DATA']:
            yield row

    def update(self, cname):
        "Update internal database with fresh snapshot of data"
        if  self.verbose:
            print "%s update %s" % (self.name, cname)
        self.storage.cleanup(cname)
        docs = self.fetch(cname)
        self.storage.insert(cname, docs)

    def dataset_stat(self, time1, time2):
        "Retrieve dataset popularity info for given time frame"
        api = 'DSStatInTimeWindow'
        params = {'tstart':time1, 'tstop':time2}
        res = self.fetch(api, params)
        for row in res:
            rec = dict(naccess=float(row['NACC']),
                    totcpu=float(row['TOTCPU']),
                    nusers=float(row['NUSERS']),
                    rnaccess=float(row['RNACC']),
                    rtotcpu=float(row['RTOTCPU']),
                    rnusers=float(row['RNUSERS']),
                    dataset=row['COLLNAME'])
            yield rec

File: Popularity/DCAFPilot/src/Go/dataframe/src/cms/popdb.go
/*
 *
 * Author     : Valentin Kuznetsov <vkuznet AT gmail dot com>
 * Description: part of cms package responsible for popDB calls
 * Created    : Wed Feb 10 19:31:44 EST 2016
 */
package cms

import (
	"encoding/json"
	"fmt"
	"strconv"
	"utils"
)

// helper function to load PopDB data stream
func loadPopDBData(furl string, data []byte) []Record {
	var out []Record
	var rec Record
	err := json.Unmarshal(data, &rec)
	if err != nil {
		msg := fmt.Sprintf("PopDB unable to unmarshal the data, furl=%s, data=%s, error=%v", furl, string(data), err)
		fmt.Println(msg)
		return out
	}
	values := rec["DATA"].([]interface{})
	for _, item := range values {
		rec := item.(map[string]interface{})
		dataset := rec["COLLNAME"].(string)
		row := make(Record)
		if datasetNameOk(dataset) {
			row["dataset"] = dataset
			row["naccess"] = rec["NACC"].(float64)
			row["totcpu"] = rec["TOTCPU"].(float64)
			row["nusers"] = rec["NUSERS"].(float64)
			row["rnaccess"], _ = strconv.ParseFloat(rec["RNACC"].(string), 10)
			row["rtotcpu"], _ = rec["RTOTCPU"].(float64)
			row["rnusers"], _ = strconv.ParseFloat(rec["RNUSERS"].(string), 10)
			out = append(out, row)
		}
	}
	return out
}

// convert YYYYDDMM into popdb notation
func popDBtstamp(ts string) string {
	return fmt.Sprintf("%s-%s-%s", ts[0:4], ts[4:6], ts[6:8])
}

// helper function to collect dataset usage from popularity DB
func datasetStats(start, stop string) []Record {
	var out []Record
	api := "DSStatInTimeWindow"
	tstart := popDBtstamp(start)
	tstop := popDBtstamp(stop)
	furl := fmt.Sprintf("%s/%s/?tstart=%s&tstop=%s", popdbUrl(), api, tstart, tstop)
	if utils.VERBOSE > 1 {
		fmt.Println("furl", furl)
	}
	response := utils.FetchResponse(furl, "")
	if response.Error == nil {
		records := loadPopDBData(furl, response.Data)
		return records
	}
	return out
}
